js_modules/dagster-ui/packages/ui-core/src/asset-selection/__tests__/AntlrAssetSelection.test.ts

```diff
@@ -99,25 +104,33 @@ describe('parseAssetSelectionQuery', () => {
     it('should parse upstream plus query', () => {
       assertQueryResult('+key:A', ['A']);
       assertQueryResult('+key:B', ['A', 'B']);
-      assertQueryResult('++key:C', ['A', 'B', 'C']);
+      assertQueryResult('+key:C', ['B', 'B2', 'C']);
+      assertQueryResult('++key:C', ['A', 'B', 'B2', 'C']);
     });
 
     it('should parse downstream plus query', () => {
+      assertQueryResult('key:A+', ['A', 'B', 'B2']);
+      assertQueryResult('key:A++', ['A', 'B', 'B2', 'C']);
       assertQueryResult('key:C+', ['C']);
       assertQueryResult('key:B+', ['B', 'C']);
-      assertQueryResult('key:A++', ['A', 'B', 'C']);
     });
 
     it('should parse upstream star query', () => {
       assertQueryResult('*key:A', ['A']);
       assertQueryResult('*key:B', ['A', 'B']);
-      assertQueryResult('**key:C', ['A', 'B', 'C']);
+      assertQueryResult('*key:C', ['A', 'B', 'B2', 'C']);
     });
 
     it('should parse downstream star query', () => {
-      assertQueryResult('key:C*', ['C']);
+      assertQueryResult('key:A*', ['A', 'B', 'B2', 'C']);
       assertQueryResult('key:B*', ['B', 'C']);
-      assertQueryResult('key:A**', ['A', 'B', 'C']);
+      assertQueryResult('key:C*', ['C']);
+    });
+
+    it('should parse up and down traversal queries', () => {
+      assertQueryResult('key:A* and *key:C', ['A', 'B', 'B2', 'C']);
+      assertQueryResult('*key:B*', ['A', 'B', 'C']);
+      assertQueryResult('key:A* and *key:C and *key:B*', ['A', 'B', 'C']);
```

Comment:
can we also add this case:

`assertQueryResult('key:A* and *key:B* and *key:C', ['A', 'B', 'C']);`

just to ensure its commutative in case somebody changes the syntax in the future